File: mobile-rn/src/screens/EditProductScreen.js
import { useEffect, useState } from "react";
import { Image, Pressable, ScrollView, StyleSheet, Text, View } from "react-native";
import * as ImagePicker from "expo-image-picker";
import { colors, radius } from "../theme";
import { Field, PrimaryButton, SecondaryButton } from "../components/ui";
import { api, ApiError } from "../api";
import { useAuth } from "../context/AuthContext";

export default function EditProductScreen({ route, navigation }) {
  const { product, onSaved } = route.params;
  const { auth, setAuth } = useAuth();

  const [categories, setCategories] = useState([]);
  const [photo, setPhoto] = useState(null);
  const [name, setName] = useState(product.name);
  const [category, setCategory] = useState(product.category || "");
  const [error, setError] = useState("");
  const [submitting, setSubmitting] = useState(false);

  const existingImageUrl = product.imageUrl.startsWith("http") ? product.imageUrl : `${api.base}${product.imageUrl}`;

  useEffect(() => {
    api.categories().then(setCategories).catch(() => {});
  }, []);

  async function pickPhoto(fromCamera) {
    const perm = fromCamera
      ? await ImagePicker.requestCameraPermissionsAsync()
      : await ImagePicker.requestMediaLibraryPermissionsAsync();
    if (!perm.granted) return;
    const result = fromCamera
      ? await ImagePicker.launchCameraAsync({ quality: 0.85 })
      : await ImagePicker.launchImageLibraryAsync({ mediaTypes: ["images"], quality: 0.85 });
    if (result.canceled) return;
    const asset = result.assets[0];
    setPhoto({ uri: asset.uri, name: asset.fileName || asset.uri.split("/").pop() || `photo-${Date.now()}.jpg`, type: asset.mimeType || "image/jpeg" });
  }

  async function handleSubmit() {
    setError("");
    if (!name.trim()) {
      setError("Le nom du produit est requis.");
      return;
    }
    setSubmitting(true);

    const fd = new FormData();
    fd.append("name", name.trim());
    fd.append("category", category);
    if (photo) fd.append("image", { uri: photo.uri, name: photo.name, type: photo.type });

    try {
      await api.updateProduct(auth.token, product.id, fd);
      onSaved?.();
      navigation.goBack();
    } catch (err) {
      if (err instanceof ApiError && err.status === 401) {
        setAuth(null);
        navigation.goBack();
        return;
      }
      setError(err instanceof ApiError ? err.message : "Une erreur s'est produite.");
    } finally {
      setSubmitting(false);
    }
  }

  return (
    <ScrollView contentContainerStyle={styles.container}>
      <Pressable onPress={() => pickPhoto(false)}>
        <Image source={{ uri: photo?.uri || existingImageUrl }} style={styles.preview} />
        <View style={styles.hint}>
          <Text style={styles.hintText}>Cliquez pour remplacer la photo</Text>
        </View>
      </Pressable>

      <View style={styles.photoRow}>
        <SecondaryButton title="📷 Prendre une photo" onPress={() => pickPhoto(true)} style={styles.flex1} />
        <SecondaryButton title="🖼️ Galerie" onPress={() => pickPhoto(false)} style={styles.flex1} />
      </View>

      <Field placeholder="Nom du produit" maxLength={80} value={name} onChangeText={setName} style={styles.gap} />

      <View style={styles.chips}>
        {categories.map((c) => (
          <Pressable key={c.name} onPress={() => setCategory(c.name)} style={[styles.chip, category === c.name && styles.chipActive]}>
            <Text style={[styles.chipText, category === c.name && styles.chipTextActive]}>{c.name}</Text>
          </Pressable>
        ))}
      </View>

      <PrimaryButton title="Enregistrer les modifications" onPress={handleSubmit} loading={submitting} />
      {!!error && <Text style={styles.error}>{error}</Text>}
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  container: { padding: 20 },
  preview: { width: "100%", aspectRatio: 1, borderRadius: radius, backgroundColor: colors.panel },
  hint: {
    position: "absolute",
    bottom: 8,
    alignSelf: "center",
    backgroundColor: "rgba(16,15,22,0.75)",
    borderRadius: 20,
    paddingHorizontal: 10,
    paddingVertical: 4
  },
  hintText: { color: colors.text, fontSize: 10 },
  photoRow: { flexDirection: "row", gap: 8, marginTop: 12, marginBottom: 16 },
  flex1: { flex: 1 },
  gap: { marginBottom: 12 },
  chips: { flexDirection: "row", flexWrap: "wrap", gap: 8, marginBottom: 16 },
  chip: { borderWidth: 1, borderColor: colors.hairline, borderRadius: 20, paddingHorizontal: 12, paddingVertical: 7 },
  chipActive: { borderColor: colors.amber, backgroundColor: "rgba(232,163,61,0.1)" },
  chipText: { color: colors.textDim, fontSize: 12 },
  chipTextActive: { color: colors.amber },
  error: { color: colors.error, fontSize: 11.5, marginTop: 10 }
});

File: mobile-flutter/lib/screens/edit_product_screen.dart
import 'dart:typed_data';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:http_parser/http_parser.dart';
import 'package:image_picker/image_picker.dart';
import 'package:provider/provider.dart';
import '../models/product.dart';
import '../services/api.dart';
import '../services/auth_provider.dart';
import '../theme/colors.dart';

class EditProductScreen extends StatefulWidget {
  final Product product;
  const EditProductScreen({super.key, required this.product});

  @override
  State<EditProductScreen> createState() => _EditProductScreenState();
}

class _EditProductScreenState extends State<EditProductScreen> {
  final _picker = ImagePicker();
  late final TextEditingController _nameCtrl = TextEditingController(text: widget.product.name);
  late String? _category = widget.product.category;

  Uint8List? _photoBytes;
  String? _photoName;
  String? _photoMime;
  List<Category> _categories = [];

  String _error = '';
  bool _submitting = false;

  @override
  void initState() {
    super.initState();
    Api.instance.categories().then((raw) {
      if (!mounted) return;
      setState(() => _categories = raw.map((e) => Category.fromJson(e as Map<String, dynamic>)).toList());
    }).catchError((_) => null);
  }

  Future<void> _pick(ImageSource source) async {
    final file = await _picker.pickImage(source: source, imageQuality: 85);
    if (file == null) return;
    final bytes = await file.readAsBytes();
    setState(() {
      _photoBytes = bytes;
      _photoName = file.name;
      _photoMime = file.mimeType;
    });
  }

  Future<void> _submit() async {
    setState(() => _error = '');
    if (_nameCtrl.text.trim().isEmpty) {
      setState(() => _error = 'Le nom du produit est requis.');
      return;
    }

    final auth = context.read<AuthProvider>().auth;
    if (auth == null) return;

    setState(() => _submitting = true);
    try {
      final image = _photoBytes != null
          ? http.MultipartFile.fromBytes(
              'image',
              _photoBytes!,
              filename: _photoName ?? 'photo.jpg',
              contentType: MediaType.parse(_photoMime ?? 'image/jpeg'),
            )
          : null;
      await Api.instance.updateProduct(auth.token, widget.product.id, _nameCtrl.text.trim(), _category ?? widget.product.category, image);
      if (mounted) Navigator.of(context).pop(true);
    } catch (err) {
      setState(() => _error = err is ApiException ? err.message : 'Une erreur s\'est produite.');
    } finally {
      if (mounted) setState(() => _submitting = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    final existingImageUrl = widget.product.imageUrl.startsWith('http') ? widget.product.imageUrl : '${Api.base}${widget.product.imageUrl}';

    return Scaffold(
      appBar: AppBar(title: const Text('Modifier l\'annonce')),
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(20),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            GestureDetector(
              onTap: () => _pick(ImageSource.gallery),
              child: Stack(
                alignment: Alignment.bottomCenter,
                children: [
                  AspectRatio(
                    aspectRatio: 1,
                    child: Container(
                      clipBehavior: Clip.antiAlias,
                      decoration: BoxDecoration(color: SnapyColors.panel, borderRadius: BorderRadius.circular(snapyRadius)),
                      child: _photoBytes != null
                          ? Image.memory(_photoBytes!, fit: BoxFit.cover)
                          : Image.network(existingImageUrl, fit: BoxFit.cover),
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.only(bottom: 8),
                    child: Container(
                      padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 4),
                      decoration: BoxDecoration(color: Colors.black.withValues(alpha: 0.6), borderRadius: BorderRadius.circular(20)),
                      child: const Text('Cliquez pour remplacer la photo', style: TextStyle(color: Colors.white, fontSize: 10)),
                    ),
                  ),
                ],
              ),
            ),
            const SizedBox(height: 12),
            Row(
              children: [
                Expanded(child: OutlinedButton(onPressed: () => _pick(ImageSource.camera), child: const Text('📷 Prendre une photo'))),
                const SizedBox(width: 8),
                Expanded(child: OutlinedButton(onPressed: () => _pick(ImageSource.gallery), child: const Text('🖼️ Galerie'))),
              ],
            ),
            const SizedBox(height: 16),
            TextField(controller: _nameCtrl, maxLength: 80, decoration: const InputDecoration(hintText: 'Nom du produit')),
            Wrap(
              spacing: 8,
              runSpacing: 8,
              children: _categories
                  .map(
                    (c) => ChoiceChip(
                      label: Text(c.name),
                      selected: _category == c.name,
                      onSelected: (_) => setState(() => _category = c.name),
                      backgroundColor: SnapyColors.panel,
                      selectedColor: const Color(0x1AE8A33D),
                      side: BorderSide(color: _category == c.name ? SnapyColors.amber : SnapyColors.hairline),
                      labelStyle: TextStyle(color: _category == c.name ? SnapyColors.amber : SnapyColors.textDim, fontSize: 12),
                    ),
                  )
                  .toList(),
            ),
            const SizedBox(height: 16),
            ElevatedButton(
              onPressed: _submitting ? null : _submit,
              child: _submitting
                  ? const SizedBox(width: 20, height: 20, child: CircularProgressIndicator(strokeWidth: 2, color: SnapyColors.amberOn))
                  : const Text('Enregistrer les modifications'),
            ),
            if (_error.isNotEmpty)
              Padding(
                padding: const EdgeInsets.only(top: 10),
                child: Text(_error, style: const TextStyle(color: SnapyColors.error, fontSize: 11.5)),
              ),
          ],
        ),
      ),
    );
  }
}
